ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,458 @@
+3.  The integration-service expects to be able to create **ephemeral** [Environments] for automated testing purposes ([STONE-114](https://issues.redhat.com/browse/STONE-114)). For our short-term goals, the automated testing use case requires the same kind of compute as for the dev and stage [Environments] (devsandbox member cluster namespaces), but will expand to include other kinds of deployment targets in the future - like hypershift clusters ([STONE-185](https://issues.redhat.com/browse/STONE-185)).
+
+## Decision
+
+We are going split the two original purposes of the [Environment] CR into different APIs.
+
+- The [Environment] will still represent the first purpose of requesting that HAS and the GitOps service **recognize a new deployment destination**.
+- For the second purpose of **provisioning a new deployment target** we are going to introduce a new [DeploymentTarget] and [DeploymentTargetClaim] API, which will be referenced from an [Environment].
+
+The new [DeploymentTarget] API is designed to emulate storage management APIs (see [persistent-volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) and [storage provisioner design](https://github.com/kubernetes-csi/external-provisioner/blob/master/doc/design.md) for reference).
+
+### CRDs
+
+#### DeploymentTarget (DT)
+
+A deployment target, usually a K8s api endpoint. The credentials for connecting to the target will
+be stored in a secret which will be referenced in the clusterCredentialsSecret field. A [DT] Can be
+created manually by a user, or dynamically using a provisioner. The `phase` section shows the lifecycle phase of the [DT].
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTarget
+metadata:
+    name: prod-dt
+spec:
+    deploymentTargetClassName: isolation-level-namespace
+    kubernetesCredentials:
+        defaultNamespace: team-a--prod-dtc
+        apiURL: …
+        clusterCredentialsSecret: team-a--prod-dtc--secret
+    claimRef:
+        name: prod-dtc
+status:
+    phase: Bound
+
+```
+
+Phases:
+
+**Pending:** DT is not yet available for binding.
+
+**Available:** DT waits for a Claim to be bound to.
+
+**Bound:** The DT was bounded to a DTC.
+
+**Released:** The DT was previously bound to a DTC which got deleted. external resources were not freed.
+
+**Failed:** DT was released from its claim, but there was a failure during the release of external resources.
+
+
+#### DeploymentTargetClaim (DTC)
+
+Represents a request for a [DeploymentTarget]. The `phase` section shows the lifecycle phase of the [DTC].
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTargetClaim
+metadata:
+    name: prod-dtc
+spec:
```

Comment:
No `selector` support as it is currently designed. As designed:

* A StorageClass references exactly one *provider*. Multiple StorageClasses can reference the same provider.
* A DeploymentTargetClaim references exactly one StorageClass, by name in the `deploymentTargetClassName`.

With selection, I suppose you could claim a target matching _any_ storage class or (as you suggest) claim a target matching any of the storage classes that provide cluster-level isolation (but not those that provide only namespace isolation).

---

In the current design, the binding controller sets an annotation on the DTC with the provisioner it found on the associated StorageClass. Provisioners only process DTCs annotated for them.

If we added a `storageClassSelector` field later in addition to the `storageClassName, we could add support for processing that selection in the binding controller. I hesitate to add it now, to avoid adding any more complexity to the initial design than necessary.